// Structure of a configuration file

syntax = "proto3";

package acppsrv.proto;

// Logging configuration
enum LogLevel {
    option allow_alias = true;
    // Levels are not the same as in log.hhp due to protobuf specifics
    unspec = 0;
    off = 1;
    X = 2;
    emerg = 2;
    emergency = 2;
    A = 3;
    alert = 3;
    C = 4;
    crit = 4;
    critical = 4;
    E = 5;
    err = 5;
    error = 5;
    W = 6;
    warning = 6;
    warn = 6;
    N = 7;
    notice = 7;
    I = 8;
    info = 8;
    D = 9;
    debug = 9;
}

message Log {
    LogLevel level = 1;
}

// Definition of a thread pool
message ThreadPool {
    // The number of threads in the pool
    int32 threads = 1;
}

// Thread pools
message ThreadPools {
    // The main server client handling pool
    ThreadPool main = 1;
    // The administrative control pool
    ThreadPool control = 2;
    // The database handling pool
    ThreadPool database = 3;
}

// HTTP server
message HttpServer {
    uint32 port = 1; // 0 = default (80)
    uint32 listen_queue = 2; // 0 = default
    uint32 max_connections = 3; // 0 = unlimited
    float idle_timeout = 4; // 0 = unlimited
    float keepalive_timeout = 5; // 0 = unlimited
    uint32 keepalive_requests = 6; // 0 = unlimited
    uint64 max_request_headers = 7; // 0 = unlimited
    uint64 max_request_body = 8; // 0 = unlimited
}

// SQLite3 database
message SQLite3Db {
    string file = 1;
    map<string, string> queries = 2;
}

message SQLite3 {
    map<string, SQLite3Db> sqlite3 = 1;
}

/*** Top level configuration node ********************************************/

message Configuration {
    Log log = 1;
    ThreadPools thread_pools = 2;
    HttpServer http_server = 3;
    SQLite3 databases = 4;
}
